proposals/p1885.md

```diff
@@ -0,0 +1,458 @@
+}
+```
+
+#### Mutable access
+
+TODO
+
+### Consumable
+
+An interface representing single-use sequential access type of container. Items
+ca be read (consumed) once, and are not expected to be accessible anymore after
+usage.
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   For each iteration, call `Consume() -> Optional(ElementType)`
+-   If the `Optional` is null, break the loop. Otherwise return the value held
+    by the Optional.
+
+#### Interface and usage
+
+```rust
+// Carbon definition
+interface Consumable() {
+    let ElementType:! Type
+    fn Consume[me: Self]() -> Optional(ElementType);
+}
+
+// Usage
+class MyContainer {
+    ...
+    impl as Consumable where .ElementType = i32 {
+        fn Consume[me: Self]() -> Optional(ElementType) {
+            ...
+        }
+    }
+}
+
+var container: MyContainer = {...};
+for (var a: auto in container) {}
+// `container` is now empty
+```
+
+#### TODO
+
+-   Size method?
+
+### Iterable
+
+The `Iterable` interface represents a generic sequential-access type of
+container. It relies on two new interfaces:
+
+-   `Iteratable`: Interface for a type that can return an iterator and be
+    iterated over
+-   `Iterator`: Interface for an iterator-like object
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   Get an iterator with `GetIterator()`
+-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
+    -   If the `Optional` is null, break the loop. Otherwise return the value
+        held by the Optional.
+
+#### Interface and usage
+
+```rust
+// Basic immutable iterator
+interface Iterator() {
+    let ElementType:! Type;
+    fn Next[me: Self]() -> Optional(ElementType);
+}
+
+// Iterable type
+interface Iterable() {
+    let IteratorType:! Iterator;
+    fn GetIterator[me: Self]() -> IteratorType;
+}
+```
+
+These interfaces can be used as such:
+
+```rust
+class MyContainer() {
+    ...
+    external impl as Iterable {
+        class IteratorType { // Associated type as member type
+            impl as Iterator where .ElementType = i32 {
+                ...
+
+                fn Next[addr me: *Self]() -> Optional(ElementType) {
+                    ...
+                    return if atEnd then .Empty else currItem;
+                }
+            }
+        }
+        fn GetIterator[me: Self]() -> IteratorType {
+            return IteratorType { ... }
+        }
+    }
+}
+```
+
+#### Iterator: `begin()` & `end()`
+
+The solution highlighted embeds the end condition in `Next()`, returns a null
+`Optional` at the last elements. This guarantees that `Iterator`s cannot be used
+improperly (code safety goal), and exposes a simple interface, but does have a
+performance cost (cost of the `Optional`).
```

Comment:
It seems like it could also have an interoperability and migration cost, because Carbon ranged-for loops won't work with C++ containers, and vice-versa.

Reply:
One option for addressing interoperability without binding ourselves to the C++ API would be to have the interop layer automatically inject adopters for language-conforming iterators. e.g., a Carbon `Iterable` could get adapted to look like a C++ iterator for C++ code, whereas a C++ iterator could have an external impl defined for Carbon code (and likely _should_ need that regardless), letting each language keep its own idioms.

(it may make sense to mark interop as an open question since it's not designed out?)

Reply:
I think it will be challenging to bridge the gap between these two APIs with adapters, because there's a substantial "impedance mismatch" between them that's hard to hide.

- With this design, a Carbon iterator basically plays the same role as a begin/end iterator _pair_ in C++, so there isn't a one-to-one relationship between the objects on either side of the adaptation. That will be a particular problem for automatic adaptation, but may also cause problems for manual adaptation.
- A C++ forward iterator (or better) must support calling `operator*` multiple times, returning the same result so long as the iterator is not advanced. To satisfy that requirement, an adaptor would have to cache the value returned from `Next`.
- C++ iterators support mutation via `operator*`. That compounds the previous problem, because mutation has to apply to the original element, not a cached copy.
- We could avoid those last two problems by saying that Carbon iterators are exposed in C++ as input iterators, but that excludes mutating use cases, and could dramatically degrade performance even for read-only cases, because many algorithms can't be implemented efficiently (or at all) in terms of the input iterator API.

Reply:
The last points definitely highlight the need for a `Get` function, in place of a simpler `Next`-only approach. I propose to move in that direction.